design-documents/storefront/pricing.md

```diff
@@ -0,0 +1,355 @@
+
+These limits mean that we can't implement `personalized pricing` feature by storing all prices in product documents.
+Even a large list of price books will be challenging for such approach. For such extreme use cases we may introduce a 
+`prices fallback` on a service which is designed to work with the large amount of prices.
+
+![Price fallback](pricing/pricing-fallback.png)
+
+Consequences:
+- Aggregation functions like facets in search service will not work properly for products with large amount of prices. Only default or limited number of prices will be available for faceting.
+- Second request obviously affects performance. A good news, performance will be affected only in queries which fetch products with large amount of prices. Other queries or slices with small products will not be affected.
+
+
+## Scenarios
+
+#### Simple
+
+- Admin sets a `base price` for the `simple` product
+- `product listing`, `PDP` and `checkout` scenarios contain the `base price`
+- Customer or guest are able to buy the product for the `base price`
+
+#### Customer specific price
+
+- Admin sets a `base price` for the `simple` product
+- Admin sets a `customer-specific price` for the same product
+- `product listing`, `PDP` and `checkout` scenarios contain `base price` for guests and `customer-specific price` for selected customer
+- Customer is able to buy the product for the `customer-specific price`
+
+#### Complex product pricing
+
+- Admin creates a `configurable` product and assign different `base prices` to variations
+- `product listing` scenario contains the minimum price of the variations
+- `PDP` scenario contains the minimum price of the variation and the price of currently selected variation
+- `product listing` scenario contains the price of currently selected variation
+- Customer or guest are able to buy the product variation for the price of selected variation
+
+#### Special prices
+
+- Admin sets a `base price` for the `simple` product
+- Admin sets a `special price` for the same product for the current date
+- `product listing`, `PDP` and `checkout` scenarios contain `special price`
+- Customers and guests are able to buy the product for the `special price`
+
+## Appendix
+
+### Price calculation algorithm for single website and customer group
+```php
+<?php
+
+$time = time(); $product = new \StdClass; $customerGroup = 12;
+
+// FIXED PRICE CALCULATIONS
+$fixedCalculator = function ($product) use ($customerGroup, $time) {
+    $specialPrice = ($product->getSpecialPriceFrom >= $time && $product->getSpecialPriceTo <= $time)
+        ? $product->getSpecialPrice
+        : 0;
+
+    $fixedPrice = min($product->getBasePrice(), $specialPrice);
+
+    $groupPrice = $product->getGroupPrice($customerGroup);
+    if ($groupPrice) {
+        $fixedPrice = ($groupPrice->getType() == 'percentage')
+            ? $product->getBasePrice() - $product->getBasePrice() / 100 * $groupPrice->getValue()
+            : $groupPrice->getValue;
```

Comment:
we do we ignore special price? I think correct behavior should be:

fixed price = customer_group_price
  ? min (special, customer_group_price)
  : min (special, base_price)